google/detectors/credentials/generic_weak_credential_detector/src/test/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/grafana/GrafanaCredentialTesterTest.java

```diff
@@ -0,0 +1,168 @@
+    NetworkService targetNetworkService =
+        NetworkService.newBuilder()
+            .setNetworkEndpoint(
+                forHostnameAndPort(mockWebServer.getHostName(), mockWebServer.getPort()))
+            .setServiceName("http")
+            .setServiceContext(grafanaServiceContext)
+            .setSoftware(Software.newBuilder().setName("http"))
+            .build();
+
+    assertThat(tester.testValidCredentials(targetNetworkService, ImmutableList.of(WEAK_CRED_1)))
+        .containsExactly(WEAK_CRED_1);
+    mockWebServer.shutdown();
+  }
+
+  @Test
+  public void detect_weakCredentialsExist_returnsAllWeakCredentials() throws Exception {
+    startMockWebServer(
+        "/",
+        Resources.toString(
+            Resources.getResource(this.getClass(), "testdata/successfulAuthdResponse.json"),
+            UTF_8));
+    NetworkService targetNetworkService =
+        NetworkService.newBuilder()
+            .setNetworkEndpoint(
+                forHostnameAndPort(mockWebServer.getHostName(), mockWebServer.getPort()))
+            .setServiceName("http")
+            .setServiceContext(grafanaServiceContext)
+            .build();
+
+    assertThat(
+            tester.testValidCredentials(
+                targetNetworkService, ImmutableList.of(WEAK_CRED_1, WEAK_CRED_2)))
+        .containsExactly(WEAK_CRED_1, WEAK_CRED_2);
+
+    mockWebServer.shutdown();
+  }
+
+  @Test
+  public void detect_noWeakCredentials_returnsNoCredentials() throws Exception {
+    startMockWebServer(
+        "/",
+        Resources.toString(
+            Resources.getResource(this.getClass(), "testdata/successfulAuthdResponse.json"),
+            UTF_8));
+    NetworkService targetNetworkService =
+        NetworkService.newBuilder()
+            .setNetworkEndpoint(
+                forHostnameAndPort(mockWebServer.getHostName(), mockWebServer.getPort()))
+            .setServiceName("http")
+            .setServiceContext(grafanaServiceContext)
+            .build();
+
+    assertThat(tester.testValidCredentials(targetNetworkService, ImmutableList.of(WRONG_CRED_1)))
+        .isEmpty();
+
+    mockWebServer.shutdown();
+  }
+
+  private void startMockWebServer(String url, String response) throws IOException {
+    mockWebServer.setDispatcher(new RespondUserInfoResponseDispatcher(response));
+    mockWebServer.start();
+    mockWebServer.url(url);
+  }
+
+  static final class RespondUserInfoResponseDispatcher extends Dispatcher {
+    private final String userInfoResponse;
+
+    RespondUserInfoResponseDispatcher(String authenticatedUserResponse) {
+      this.userInfoResponse = checkNotNull(authenticatedUserResponse);
+    }
+
+    @Override
+    public MockResponse dispatch(RecordedRequest recordedRequest) {
+      if (recordedRequest.getPath().startsWith("/api/user")
+          && (recordedRequest.getHeaders().toString().contains(WEAK_CRED_AUTH_1)
+              || recordedRequest.getHeaders().toString().contains(WEAK_CRED_AUTH_2))) {
```

Comment:
On my side this test file does not work in the current state because `recordedRequest.getHeaders()` will return a map. The `authorization` key should be retrieved before matching the password. Here is a suggested change:

```java
      var isUserEndpoint = recordedRequest.getPath().startsWith("/api/user");
      var authHeader = recordedRequest.getHeaders().get("Authorization").toString();
      var hasWeakCred1 = authHeader.contains(WEAK_CRED_AUTH_1);
      var hasWeakCred2 = authHeader.contains(WEAK_CRED_AUTH_2);

      if (isUserEndpoint && (hasWeakCred1 || hasWeakCred2)) {
        return new MockResponse().setResponseCode(HttpStatus.OK.code()).setBody(userInfoResponse);
      }
      return new MockResponse().setResponseCode(HttpStatus.UNAUTHORIZED.code());
```

Note that it also requires changing `WEAK_CRED_AUTH_[12]`:

```java
  private static final String WEAK_CRED_AUTH_1 = "Basic dXNlcjoxMjM0";
  private static final String WEAK_CRED_AUTH_2 = "Basic cm9vdDpwYXNz";
```

Reply:
Fix in 3cf4558d4b1996a807a0efea649b9abe783fe64c